### PR TITLE
[GenSync] Introduce GenSync Interface

### PR DESCRIPTION
**Pull request by Bowenislandsong**
The GenSync interface is used as a framework to sync with distributed hosts. It supports continuous sync, and with multiple nodes. This commit contains a TCP connection interface and full sync protocol. It also updates the hash function and refactors the entire repo.

The TCP protocol is used to transfer data between nodes and the full sync is to sync data set between local and remote nodes.

### CODE REVIEW

File: pkg/lib/algorithm/iblt/sync.go [hunk 1]
```diff
@@ -1,14 +1,41 @@
...
+
+func NewIBLTSetSync(diffNum, dataLen int) (genSync.GenSync, error) {
+	if diffNum <= 0 {
+		return nil, fmt.Errorf("number of difference should be positive")
+	}
+	// TODO: table size should be 1.5 times the sym set diff, IBLT algo impl should improve.
+	tableSize := diffNum * 4
+	if tableSize < 4 {
+		tableSize = 4
+	}
```
Comment (wfystx) [inline pkg/lib/algorithm/iblt/sync.go:31]: why is this necessary?

Comment (Bowenislandsong) [reply pkg/lib/algorithm/iblt/sync.go:31]: Good question. The IBLT table size is (set to) linear to the number of symmetrical set differences between the sets. Theoretically the factor should be 1.5 but we have 4 here, I think I found the reason and will update to see if it can be cut down to a smaller number than 4. Anyway, this is to say that the minimum table size should be 4. It is a “magical” number but this number should be at least 2 for resolving 1 diff given that the minimum number of hash fxns should be 2. The hash fxn is to assign where the value sits in the table. The minimum has to be 2 to enable peeling process.
